File: Linear_Control_Course/Chapter28/Lesson5/Chapter28_Lesson5.m
% config.m (reproducible configuration)
cfg.experiment_name = 'pi_design_second_order';
cfg.plant.num = 1;
cfg.plant.den = [1 2 1];  % (s + 1)^2
cfg.controller.Kp = 4.0;
cfg.controller.Ki = 3.0;
cfg.sim.t_final = 10;
cfg.sim.n_points = 2000;
cfg.tolerances.settling_band = 0.02;

% plant and controller
s = tf('s');
G = tf(cfg.plant.num, cfg.plant.den);
C = cfg.controller.Kp + cfg.controller.Ki / s;

L = C * G;
T = feedback(L, 1);  % unity feedback

% Bode and step plots (saved to files, no manual export)
fig1 = figure('Visible', 'off');
bode(L);
grid on;
title('Open-loop Bode plot');
saveas(fig1, 'matlab_pi_openloop_bode.png');

fig2 = figure('Visible', 'off');
[y, t] = step(T, linspace(0, cfg.sim.t_final, cfg.sim.n_points));
plot(t, y); hold on;
yline(1, '--');
xlabel('Time (s)');
ylabel('Output y(t)');
title('Closed-loop step response (PI)');
grid on;
saveas(fig2, 'matlab_pi_step_response.png');

close(fig1); close(fig2);

% Compute metrics
y_final = y(end);
Mp = max(y) - 1;
ess = 1 - y_final;

idx_last_outside = find(abs(y - 1) > cfg.tolerances.settling_band, 1, 'last');
if isempty(idx_last_outside)
    ts = 0;
else
    ts = t(idx_last_outside);
end

metrics = table(Mp, ess, ts, ...
    'VariableNames', {'Overshoot', 'SteadyStateError', 'SettlingTime'});

writetable(metrics, 'matlab_pi_metrics.csv');

% Example robotics extension:
% If G is derived from a linearized robot joint model via
% linmod('robot_joint_model') or linearize from Robotics System Toolbox,
% the exact same structure applies and metrics remain reproducible.

File: Linear_Control_Course/Chapter28/Lesson5/Chapter28_Lesson5.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

class Config {
    public double Kp = 4.0;
    public double Ki = 3.0;
    public double dt = 0.001;
    public double tFinal = 10.0;
    public double settlingBand = 0.02;
}

// Same plant as before: G(s) = 1 / (s + 1)^2, PI controller
class PlantPI {
    private Config cfg;
    private double x1 = 0.0;
    private double x2 = 0.0;
    private double xI = 0.0;

    public PlantPI(Config cfg) {
        this.cfg = cfg;
    }

    public void reset() {
        x1 = 0.0;
        x2 = 0.0;
        xI = 0.0;
    }

    public double step(double r) {
        double y = x1;
        double e = r - y;
        xI += e * cfg.dt;
        double u = cfg.Kp * e + cfg.Ki * xI;

        double x1dot = -x1 + x2;
        double x2dot = -x2 + u;

        x1 += x1dot * cfg.dt;
        x2 += x2dot * cfg.dt;
        return y;
    }
}

public class PiExperiment {
    public static void main(String[] args) throws IOException {
        Config cfg = new Config();
        PlantPI system = new PlantPI(cfg);

        int nSteps = (int) Math.round(cfg.tFinal / cfg.dt);
        List<Double> tList = new ArrayList<>(nSteps + 1);
        List<Double> yList = new ArrayList<>(nSteps + 1);

        double t = 0.0;
        for (int k = 0; k <= nSteps; ++k) {
            double y = system.step(1.0);
            tList.add(t);
            yList.add(y);
            t += cfg.dt;
        }

        // Metrics
        double yFinal = yList.get(yList.size() - 1);
        double Mp = 0.0;
        for (double y : yList) {
            double val = y - 1.0;
            if (val > Mp) Mp = val;
        }
        double ess = 1.0 - yFinal;

        double ts = 0.0;
        for (int k = nSteps; k >= 0; --k) {
            double y = yList.get(k);
            if (Math.abs(y - 1.0) > cfg.settlingBand) {
                ts = tList.get(k);
                break;
            }
        }

        // CSV export
        try (FileWriter fw = new FileWriter("java_pi_closed_loop.csv")) {
            fw.write("t,y\n");
            for (int k = 0; k < tList.size(); ++k) {
                fw.write(tList.get(k) + "," + yList.get(k) + "\n");
            }
        }

        // Metrics export
        try (FileWriter fw = new FileWriter("java_pi_metrics.txt")) {
            fw.write("Kp=" + cfg.Kp + "\n");
            fw.write("Ki=" + cfg.Ki + "\n");
            fw.write("Mp=" + Mp + "\n");
            fw.write("ess=" + ess + "\n");
            fw.write("ts=" + ts + "\n");
        }

        // In robot control libraries such as WPILib, the same numerical
        // experiment can be run in a test or simulation mode, reading cfg
        // from a configuration file and writing back metrics for documentation.
    }
}
